src/solver/optimisation/constraint_builder.h

```diff
@@ -0,0 +1,376 @@
+    }
+
+    int NTCDirect(unsigned int index) const
+    {
+        return nativeOptimVar.NumeroDeVariableDeLInterconnexion[index];
+    }
+
+    int IntercoDirectCost(unsigned int index) const
+    {
+        return nativeOptimVar.NumeroDeVariableCoutOrigineVersExtremiteDeLInterconnexion[index];
+    }
+
+    int IntercoIndirectCost(unsigned int index) const
+    {
+        return nativeOptimVar.NumeroDeVariableCoutExtremiteVersOrigineDeLInterconnexion[index];
+    }
+
+    int ShortTermStorageInjection(unsigned int index) const
+    {
+        return nativeOptimVar.SIM_ShortTermStorage.InjectionVariable[index];
+    }
+
+    int ShortTermStorageWithdrawal(unsigned int index) const
+    {
+        return nativeOptimVar.SIM_ShortTermStorage.WithdrawalVariable[index];
+    }
+
+    int ShortTermStorageLevel(unsigned int index) const
+    {
+        return nativeOptimVar.SIM_ShortTermStorage.LevelVariable[index];
+    }
+
+    int HydProd(unsigned int index) const
+    {
+        return nativeOptimVar.NumeroDeVariablesDeLaProdHyd[index];
+    }
+
+    int HydProdDown(unsigned int index) const
+    {
+        return nativeOptimVar.NumeroDeVariablesVariationHydALaBaisse[index];
+    }
+
+    int HydProdUp(unsigned int index) const
+    {
+        return nativeOptimVar.NumeroDeVariablesVariationHydALaHausse[index];
+    }
+
+    int Pumping(unsigned int index) const
+    {
+        return nativeOptimVar.NumeroDeVariablesDePompage[index];
+    }
+
+    int HydroLevel(unsigned int index) const
+    {
+        return nativeOptimVar.NumeroDeVariablesDeNiveau[index];
+    }
+
+    int Overflow(unsigned int index) const
+    {
+        return nativeOptimVar.NumeroDeVariablesDeDebordement[index];
+    }
+
+    int FinalStorage(unsigned int index) const
+    {
+        return NumeroDeVariableStockFinal[index];
+    }
+
+    int LayerStorage(unsigned area, unsigned layer) const
+    {
+        return NumeroDeVariableDeTrancheDeStock[area][layer];
+    }
+
+    int PositiveUnsuppliedEnergy(unsigned int index) const
+    {
+        return nativeOptimVar.NumeroDeVariableDefaillancePositive[index];
+    }
+
+    int NegativeUnsuppliedEnergy(unsigned int index) const
+    {
+        return nativeOptimVar.NumeroDeVariableDefaillanceNegative[index];
+    }
+
+private:
+    const CORRESPONDANCES_DES_VARIABLES& nativeOptimVar;
+    const std::vector<int>& NumeroDeVariableStockFinal;
+    const std::vector<std::vector<int>>& NumeroDeVariableDeTrancheDeStock;
+};
+
+} // namespace Variable
+
+class ConstraintBuilder
+{
+public:
+    ConstraintBuilder(
+      PROBLEME_HEBDO& problemeHebdo,
+      const std::vector<CORRESPONDANCES_DES_VARIABLES>& CorrespondanceVarNativesVarOptim) :
+     problemeHebdo(problemeHebdo),
+     problemeAResoudre(*problemeHebdo.ProblemeAResoudre),
+     varNative(CorrespondanceVarNativesVarOptim)
+    {
+    }
+
+    ConstraintBuilder& updateHourWithinWeek(unsigned hour)
```

Comment:
This method seems dangerous and seems to overlap with the role of `shift` argument in the other methods.

To clarify that, I would propose that the "base hour" for a given constraint is **never** changed while building a constraint. Instead, we should use shift arguments in those places. Otherwise, one could make a mistake by changing involuntarily the base hour for subsequent calls.

To enforce this, the best design would be that a constraint builder is used only for creating one constraint, and its base hour is defined in its constructor and not changeable afterwards.
Note that otherwise, another risk is that we could re-use the builder and forget to change its base hour.

Reply:
i'm agree with you on building one constraint at time, unfortunately we have some weird timestep stuff like in [src/solver/optimisation/constraints/MinDownTime.cpp](https://github.com/AntaresSimulatorTeam/Antares_Simulator/blob/31b39eface118aa51846ea816c80f44ffec43973/src/solver/optimisation/constraints/MinDownTime.cpp)
and it has to be addressed in dedicated Pr